### PR TITLE
rosbag2 splitting changes part 1

### PR DESCRIPTION
**Pull request by prajakta-gokhale**
This is the first part of changes for rosbag2 splitting.

This includes:
* `ros2 bag record` has a new CLI option to specify bagfile-size
* bag-size is passed through python bindings and stored as a `storage_option`
* bag-size value is passed to storage plugin interfaces to use it in the plugins (plugin changes will be in the next part)

### CODE REVIEW

File: ros2bag/ros2bag/verb/record.py [hunk 1]
```diff
@@ -52,6 +51,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             help='time in ms to wait between querying available topics for recording. It has no '
              'effect if --no-discovery is enabled.'
         )
+        parser.add_argument(
+            '-b', '--bag-size', type=int, default=sys.maxsize,
```
Comment (dabonnie) [inline ros2bag/ros2bag/verb/record.py]: What are the units of the size?

Comment (prajakta-gokhale) [reply ros2bag/ros2bag/verb/record.py]: It's in `bytes`, I'll add that in the `help` section.

Comment (dabonnie) [reply ros2bag/ros2bag/verb/record.py]: nit: I'd include the size in the variable name so intent is clear in code as well.

**Review by dabonnie — APPROVED** [review]

File: rosbag2/include/rosbag2/storage_options.hpp [hunk 2]
```diff
@@ -25,6 +25,7 @@ struct StorageOptions
 public:
   std::string uri;
   std::string storage_id;
+  uint64_t max_bagfile_size = UINT64_MAX;
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/storage_options.hpp]: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-in-class-initializer

File: ros2bag/ros2bag/verb/record.py [hunk 3]
```diff
@@ -52,6 +51,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
...
+            help='provides a threshold for recording bagfile size in bytes, before the recording is split and '
+              'rolled over to a new bagfile. This value is only used as a guideline because actually enforcing a '
+              'maximum file size is difficult. The actual bagfile size may be a little over this value. '
+              'Defaults to system maximum integer value.'
```
Comment (thomas-moulard) [inline ros2bag/ros2bag/verb/record.py]: `Defaults to system maximum integer value.` is not understandable by users - could we make it "By default, all data are stored in a single file and this feature is disabled."

File: ros2bag/ros2bag/verb/record.py (hunk 1 above)
Comment (thomas-moulard) [inline ros2bag/ros2bag/verb/record.py]: There are multiple defaults values through the code: `sys.maxsize` there, then `UINT64_MAX`. Let's use '0' as the default and make this mean "no splitting".

File: rosbag2/include/rosbag2/storage_options.hpp (hunk 2 above)
Comment (thomas-moulard) [reply rosbag2/include/rosbag2/storage_options.hpp]: By using default initialization, I believe this is not POD anymore, the other files should have defaults too.

File: rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp [hunk 4]
```diff
@@ -46,22 +47,25 @@ rosbag2_transport_record(PyObject * Py_UNUSED(self), PyObject * args, PyObject *
   bool all = false;
   bool no_discovery = false;
   uint64_t polling_interval_ms = 100;
+  uint64_t max_bagfile_size = UINT64_MAX;
```
Comment (thomas-moulard) [inline rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp]: This does not match the doc:
https://docs.python.org/3/c-api/arg.html#numbers

It should be an `unsigned long long` and not an `uint64_t`. The current code is also wrong on 32-bits arch. See `stdint.h`:

```c++
#if __WORDSIZE == 64
typedef unsigned long int   uint64_t;
#else
__extension__
typedef unsigned long long int  uint64_t;
#endif
```

**Comment by prajakta-gokhale** [comment]
Changes in review in upstream repo.

**Pull request closed without merging**